crates/tx5-go-pion/src/data_chan.rs

```diff
@@ -1,106 +1,209 @@
-    {
-        let cb: DataChanEvtCb = Arc::new(cb);
-        let data_chan_id = self.0;
-        self.0 = 0;
-        replace_data_chan_evt_cb(data_chan_id, move || {
-            for evt in self.1.lock().drain(..) {
-                cb(evt);
+    pub fn close(&mut self, err: Error) {
+        // self.evt_send.send_err() is called in Drop impl
+        self.drop_err = err;
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct WeakDataChan(
+    pub(crate) Weak<Mutex<std::result::Result<DataChanCore, Error>>>,
+);
+
+macro_rules! data_chan_strong_core {
+    ($inner:expr, $ident:ident, $block:block) => {
+        match &mut *$inner.lock().unwrap() {
+            Ok($ident) => $block,
+            Err(err) => Result::Err(err.clone().into()),
+        }
+    };
+}
+
+macro_rules! data_chan_weak_core {
+    ($inner:expr, $ident:ident, $block:block) => {
+        match $inner.upgrade() {
+            Some(strong) => data_chan_strong_core!(strong, $ident, $block),
+            None => Result::Err(Error::id("DataChannelClosed")),
+        }
+    };
+}
+
+impl WeakDataChan {
+    pub fn close(&self, err: Error) {
+        if let Some(strong) = self.0.upgrade() {
+            let mut tmp = Err(err.clone());
+
+            {
+                let mut lock = strong.lock().unwrap();
+                if let Ok(core) = &mut *lock {
+                    core.close(err);
+                }
+                std::mem::swap(&mut *lock, &mut tmp);
             }
-            cb
-        });
-        DataChannel(data_chan_id)
+
+            // make sure the above lock is released before this is dropped
+            drop(tmp);
+        }
+    }
+
+    pub fn get_recv_limit(&self) -> Result<Arc<tokio::sync::Semaphore>> {
+        data_chan_weak_core!(self.0, core, { Ok(core.recv_limit.clone()) })
+    }
+
+    pub async fn send_evt(&self, evt: DataChannelEvent) -> Result<()> {
+        data_chan_weak_core!(self.0, core, { Ok(core.evt_send.clone()) })?
+            .send(evt)
+            .await
     }
 }
 
 /// A go pion webrtc DataChannel.
-#[derive(Debug)]
-pub struct DataChannel(pub(crate) usize);
+pub struct DataChannel(Arc<Mutex<std::result::Result<DataChanCore, Error>>>);
 
-impl Drop for DataChannel {
-    fn drop(&mut self) {
-        unregister_data_chan_evt_cb(self.0);
-        unsafe { API.data_chan_free(self.0) }
+impl std::fmt::Debug for DataChannel {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let res = match &*self.0.lock().unwrap() {
+            Ok(core) => format!("DataChannel(open, {})", core.data_chan_id),
+            Err(err) => format!("DataChannel(closed, {:?})", err),
+        };
+        f.write_str(&res)
     }
 }
 
 impl DataChannel {
+    pub(crate) fn new(
+        data_chan_id: usize,
+        recv_limit: Arc<tokio::sync::Semaphore>,
+    ) -> (Self, EventRecv<DataChannelEvent>) {
+        let (evt_send, evt_recv) = EventSend::new(1024);
+
+        let strong = Arc::new(Mutex::new(Ok(DataChanCore::new(
```

Comment:
How about this at least?

```suggestion

    const CAPACITY: usize = 1024;

    pub(crate) fn new(
        data_chan_id: usize,
        recv_limit: Arc<tokio::sync::Semaphore>,
    ) -> (Self, EventRecv<DataChannelEvent>) {
        let (evt_send, evt_recv) = EventSend::new(Self::CAPACITY);

        let strong = Arc::new(Mutex::new(Ok(DataChanCore::new(
```

Reply:
This has been made into a config field.